# Importar librerías

In [1]:
import pandas as pd
import a_funciones as funciones
import pyarrow # para guardar en formato .parquet

8


# Carga de datos

In [2]:
circuits = pd.read_csv('../data/circuits.csv')
constructors =  pd.read_csv('../data/constructors.csv')
drivers = pd.read_csv('../data/drivers.csv')
lap_times = pd.read_csv('../data/lap_times.csv')
pit_stop = pd.read_csv('../data/pit_stops.csv')
races = pd.read_csv('../data/races.csv')
results = pd.read_csv('../data/results.csv')
status = pd.read_csv('../data/status.csv')

# Análisis de los DF

## Circuits

In [41]:
circuits.head()

circuitId   circuitRef                            name      location  \
0          1  albert_park  Albert Park Grand Prix Circuit     Melbourne   
1          2       sepang    Sepang International Circuit  Kuala Lumpur   
2          3      bahrain   Bahrain International Circuit        Sakhir   
3          4    catalunya  Circuit de Barcelona-Catalunya      Montmeló   
4          5     istanbul                   Istanbul Park      Istanbul   

     country       lat        lng  alt  \
0  Australia -37.84970  144.96800   10   
1   Malaysia   2.76083  101.73800   18   
2    Bahrain  26.03250   50.51060    7   
3      Spain  41.57000    2.26111  109   
4     Turkey  40.95170   29.40500  130   

                                                 url  
0  http://en.wikipedia.org/wiki/Melbourne_Grand_P...  
1  http://en.wikipedia.org/wiki/Sepang_Internatio...  
2  http://en.wikipedia.org/wiki/Bahrain_Internati...  
3  http://en.wikipedia.org/wiki/Circuit_de_Barcel...  
4         http://en.wikipedia.org/wiki/Istanbul_Park

In [42]:
circuits.tail()

circuitId circuitRef                                  name   location  \
72         75   portimao    Autódromo Internacional do Algarve   Portimão   
73         76    mugello  Autodromo Internazionale del Mugello    Mugello   
74         77     jeddah               Jeddah Corniche Circuit     Jeddah   
75         78     losail          Losail International Circuit  Al Daayen   
76         79      miami         Miami International Autodrome      Miami   

         country      lat      lng  alt  \
72      Portugal  37.2270  -8.6267  108   
73         Italy  43.9975  11.3719  255   
74  Saudi Arabia  21.6319  39.1044   15   
75         Qatar  25.4900  51.4542   \N   
76           USA  25.9581 -80.2389   \N   

                                                  url  
72  http://en.wikipedia.org/wiki/Algarve_Internati...  
73       http://en.wikipedia.org/wiki/Mugello_Circuit  
74  http://en.wikipedia.org/wiki/Jeddah_Street_Cir...  
75  http://en.wikipedia.org/wiki/Losail_Internatio...  
76  http://en.wikipedia.org/wiki/Miami_Internation...

In [43]:
# Elijo las columnas que son relevantes para el análisis
circuits_cols = circuits[["circuitId","location","country","lat","lng"]]

In [44]:
# Se revisa la dimensión, valores nulos, duplicados y los tipos de datos.
funciones.nudu(circuits_cols)

Dimensión del DF: (77, 5) 

Valores nulos por columna:
circuitId    0
location     0
country      0
lat          0
lng          0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
circuitId      int64
location      object
country       object
lat          float64
lng          float64
dtype: object


In [45]:
funciones.values_unique(circuits_cols)

Valores únicos en la columna circuitId:
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 80 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 73
 75 76 77 78 79]

Valores únicos en la columna location:
 ['Melbourne' 'Kuala Lumpur' 'Sakhir' 'Montmeló' 'Istanbul' 'Monte-Carlo'
 'Montreal' 'Magny Cours' 'Silverstone' 'Hockenheim' 'Budapest' 'Valencia'
 'Spa' 'Monza' 'Marina Bay' 'Oyama' 'Shanghai' 'São Paulo' 'Indianapolis'
 'Nürburg' 'Imola' 'Suzuka' 'Las Vegas' 'Abu Dhabi' 'Buenos Aires'
 'Jerez de la Frontera' 'Estoril' 'Okayama' 'Adelaide' 'Midrand'
 'Castle Donington' 'Mexico City' 'Phoenix' 'Le Castellet'
 'Yeongam County' 'Rio de Janeiro' 'Detroit' 'Kent' 'Zandvoort'
 'Heusden-Zolder' 'Dijon' 'Dallas' 'California' 'Nevada' 'Madrid'
 'New York State' 'Anderstorp' 'Ontario' 'Barcelona' 'Brussels'
 'Clermont-Ferrand' 'Quebec' 'Rouen' 'Le Mans' 'Reims'
 'Eastern Cape Provi

## Constructors

In [46]:
constructors.head()

constructorId constructorRef        name nationality  \
0              1        mclaren     McLaren     British   
1              2     bmw_sauber  BMW Sauber      German   
2              3       williams    Williams     British   
3              4        renault     Renault      French   
4              5     toro_rosso  Toro Rosso     Italian   

                                                 url  
0               http://en.wikipedia.org/wiki/McLaren  
1            http://en.wikipedia.org/wiki/BMW_Sauber  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  
3  http://en.wikipedia.org/wiki/Renault_in_Formul...  
4   http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso

In [47]:
constructors.tail()

constructorId constructorRef            name nationality  \
206            209          manor  Manor Marussia     British   
207            210           haas    Haas F1 Team    American   
208            211   racing_point    Racing Point     British   
209            213     alphatauri      AlphaTauri     Italian   
210            214         alpine  Alpine F1 Team      French   

                                                   url  
206      http://en.wikipedia.org/wiki/Manor_Motorsport  
207          http://en.wikipedia.org/wiki/Haas_F1_Team  
208  http://en.wikipedia.org/wiki/Racing_Point_F1_Team  
209   http://en.wikipedia.org/wiki/Scuderia_AlphaTauri  
210        http://en.wikipedia.org/wiki/Alpine_F1_Team

In [48]:
# Elijo las columnas que son relevantes para el análisis
constructors_cols = constructors[["constructorId","name","nationality"]]

In [49]:
funciones.nudu(constructors_cols)

Dimensión del DF: (211, 3) 

Valores nulos por columna:
constructorId    0
name             0
nationality      0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
constructorId     int64
name             object
nationality      object
dtype: object


In [50]:
funciones.values_unique(constructors_cols)

Valores únicos en la columna constructorId:
 [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  44  45  46  47  48  49  50  51  52  53  54  55
  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73
  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90  91
  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108 109
 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127
 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145
 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163
 164 167 166 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182
 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200
 201 202 203 204 205 206 207 208 209 210 211 213 214]

Valores únicos en la columna name:
 ['McLaren' 'BMW Sauber' 'Williams' 'Renault' 'Toro Rosso' 'Fe

## Drivers

In [51]:
drivers.head()

driverId   driverRef number code  forename     surname         dob  \
0         1    hamilton     44  HAM     Lewis    Hamilton  1985-01-07   
1         2    heidfeld     \N  HEI      Nick    Heidfeld  1977-05-10   
2         3     rosberg      6  ROS      Nico     Rosberg  1985-06-27   
3         4      alonso     14  ALO  Fernando      Alonso  1981-07-29   
4         5  kovalainen     \N  KOV    Heikki  Kovalainen  1981-10-19   

  nationality                                             url  
0     British     http://en.wikipedia.org/wiki/Lewis_Hamilton  
1      German      http://en.wikipedia.org/wiki/Nick_Heidfeld  
2      German       http://en.wikipedia.org/wiki/Nico_Rosberg  
3     Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso  
4     Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen

In [52]:
drivers.tail()

driverId        driverRef number code forename     surname         dob  \
852       854  mick_schumacher     47  MSC     Mick  Schumacher  1999-03-22   
853       855             zhou     24  ZHO   Guanyu        Zhou  1999-05-30   
854       856         de_vries     21  DEV     Nyck    de Vries  1995-02-06   
855       857          piastri     81  PIA    Oscar     Piastri  2001-04-06   
856       858         sargeant      2  SAR    Logan    Sargeant  2000-12-31   

    nationality                                           url  
852      German  http://en.wikipedia.org/wiki/Mick_Schumacher  
853     Chinese      http://en.wikipedia.org/wiki/Zhou_Guanyu  
854       Dutch    http://en.wikipedia.org/wiki/Nyck_de_Vries  
855  Australian    http://en.wikipedia.org/wiki/Oscar_Piastri  
856    American   http://en.wikipedia.org/wiki/Logan_Sargeant

In [53]:
# Elijo las columnas que son relevantes number el análisis
drivers_cols = drivers[["driverId","forename","surname","dob","nationality"]]

# Se renombra una columna
drivers_cols = drivers_cols.rename(columns = {"dob":"birthdate"})

In [54]:
# Se revisa la dimensión, valores nulos, duplicados y los tipos de datos.
funciones.nudu(drivers_cols)

Dimensión del DF: (857, 5) 

Valores nulos por columna:
driverId       0
forename       0
surname        0
birthdate      0
nationality    0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
driverId        int64
forename       object
surname        object
birthdate      object
nationality    object
dtype: object


In [55]:
# La columna birthdate se convierte en tipo fecha
drivers_cols.birthdate = pd.to_datetime(drivers_cols.birthdate, format='%Y-%m-%d')

In [56]:
funciones.nudu(drivers_cols)

Dimensión del DF: (857, 5) 

Valores nulos por columna:
driverId       0
forename       0
surname        0
birthdate      0
nationality    0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
driverId                int64
forename               object
surname                object
birthdate      datetime64[ns]
nationality            object
dtype: object


In [57]:
funciones.values_unique(constructors_cols)

Valores únicos en la columna constructorId:
 [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  44  45  46  47  48  49  50  51  52  53  54  55
  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73
  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90  91
  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108 109
 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127
 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145
 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163
 164 167 166 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182
 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200
 201 202 203 204 205 206 207 208 209 210 211 213 214]

Valores únicos en la columna name:
 ['McLaren' 'BMW Sauber' 'Williams' 'Renault' 'Toro Rosso' 'Fe

## Lap Time

In [58]:
lap_times.head()

raceId  driverId  lap  position      time  milliseconds
0     841        20    1         1  1:38.109         98109
1     841        20    2         1  1:33.006         93006
2     841        20    3         1  1:32.713         92713
3     841        20    4         1  1:32.803         92803
4     841        20    5         1  1:32.342         92342

In [59]:
lap_times.tail()

raceId  driverId  lap  position      time  milliseconds
551737    1110       817   40        17  1:54.361        114361
551738    1110       817   41        17  1:53.367        113367
551739    1110       817   42        16  1:55.247        115247
551740    1110       817   43        16  1:52.115        112115
551741    1110       817   44        16  1:52.265        112265

In [60]:
# Elijo las columnas que son relevantes number el análisis
lap_times_cols = lap_times[["raceId","driverId","lap","position","time","milliseconds"]]

In [61]:
# Se revisa la dimensión, valores nulos, duplicados y los tipos de datos.
funciones.nudu(lap_times_cols)

Dimensión del DF: (551742, 6) 

Valores nulos por columna:
raceId          0
driverId        0
lap             0
position        0
time            0
milliseconds    0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
raceId           int64
driverId         int64
lap              int64
position         int64
time            object
milliseconds     int64
dtype: object


In [62]:
funciones.values_unique(lap_times_cols)

Valores únicos en la columna raceId:
 [ 841  842  843  844  845  846  847  848  849  850  851  852  853  854
  855  856  857  858  859  860  861  862  863  864  865  866  867  868
  869  870  871  872  873  874  875  876  877  878  879  881  880  882
  883  884  885  886  887  888  890  891  892  893  894  895  896  897
  898  899  900  901  902  903  904  905  906  907  908  909  910  911
  912  913  914  915  916  926  918  927  928  929  930  931  932  933
  934  936  937  938  939  940  941  942  943  944  945  948  949  224
  225  226  227  228  229  230  231  232  233  234  235  236  237  238
  239  207  208  209  210  211  212  213  214  215  216  217  218  219
  220  221  222  223  191  192  193  194  195  196  197  198  199  200
  201  202  203  204  205  206  175  176  177  178  179  180  181  182
  183  184  185  186  187  188  189  190  158  159  160  161  162  163
  164  165  166  167  168  169  170  171  172  173  174  141  142  143
  144  145  146  147  148  149  150  15

## Pit Stop

In [63]:
pit_stop.head()

raceId  driverId  stop  lap      time duration  milliseconds
0     841       153     1    1  17:05:23   26.898         26898
1     841        30     1    1  17:05:52   25.021         25021
2     841        17     1   11  17:20:48   23.426         23426
3     841         4     1   12  17:22:34   23.251         23251
4     841        13     1   13  17:24:10   23.842         23842

In [64]:
pit_stop.tail()

raceId  driverId  stop  lap      time duration  milliseconds
10084    1110         4     2   29  15:59:01   23.798         23798
10085    1110       830     2   30  16:00:16   23.012         23012
10086    1110       848     3   33  16:07:06   23.529         23529
10087    1110       858     3   34  16:09:09   23.109         23109
10088    1110         1     3   42  16:23:05   23.156         23156

In [65]:
# Elijo las columnas que son relevantes number el análisis
pit_stop_cols = pit_stop[["raceId","driverId","stop","lap","duration"]]

In [66]:
# Se revisa la dimensión, valores nulos, duplicados y los tipos de datos.
funciones.nudu(pit_stop_cols)

Dimensión del DF: (10089, 5) 

Valores nulos por columna:
raceId      0
driverId    0
stop        0
lap         0
duration    0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
raceId       int64
driverId     int64
stop         int64
lap          int64
duration    object
dtype: object


In [67]:
funciones.values_unique(pit_stop_cols)

Valores únicos en la columna raceId:
 [ 841  842  843  844  845  846  847  848  849  850  851  852  853  854
  855  856  857  858  859  860  861  862  863  864  865  866  867  868
  869  870  871  872  873  874  875  876  877  878  879  880  881  882
  883  884  885  886  887  888  890  891  892  893  894  895  896  897
  898  899  900  901  902  903  904  905  906  907  908  909  910  911
  912  913  914  915  916  917  918  926  927  928  929  930  931  932
  933  934  936  937  938  939  940  941  942  943  944  945  948  949
  950  951  952  953  954  955  956  957  958  959  960  961  962  963
  964  965  966  967  968  969  970  971  972  973  974  975  976  977
  978  979  980  981  982  983  984  985  986  987  988  989  990  991
  992  993  994  995  996  997  998  999 1000 1001 1002 1003 1004 1005
 1006 1007 1008 1009 1010 1011 1012 1013 1014 1015 1016 1017 1018 1019
 1020 1021 1022 1023 1024 1025 1026 1027 1028 1029 1030 1031 1032 1033
 1034 1035 1036 1037 1038 1039 1040 104

## Races

In [68]:
races.head()

raceId  year  round  circuitId                   name        date  \
0       1  2009      1          1  Australian Grand Prix  2009-03-29   
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url fp1_date  \
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...       \N   
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...       \N   
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...       \N   
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...       \N   
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...       \N   

  fp1_time fp2_date fp2_time fp3_date fp3_time quali_date quali_time  \
0       \N       \N       \N       \N       \N         \N         \N   
1       \N       \N       \N       \N       \N         \N         \N   
2       \N       \N       \N       \N       \N         \N         \N   
3       \N       \N       \N       \N       \N         \N         \N   
4       \N       \N       \N       \N       \N         \N         \N   

  sprint_date sprint_time  
0          \N          \N  
1          \N          \N  
2          \N          \N  
3          \N          \N  
4          \N          \N

In [69]:
races.tail()

raceId  year  round  circuitId                      name        date  \
1096    1116  2023     18         69  United States Grand Prix  2023-10-22   
1097    1117  2023     19         32    Mexico City Grand Prix  2023-10-29   
1098    1118  2023     20         18      São Paulo Grand Prix  2023-11-05   
1099    1119  2023     21         80      Las Vegas Grand Prix  2023-11-19   
1100    1120  2023     22         24      Abu Dhabi Grand Prix  2023-11-26   

          time                                                url    fp1_date  \
1096  19:00:00  https://en.wikipedia.org/wiki/2023_United_Stat...  2023-10-20   
1097  20:00:00  https://en.wikipedia.org/wiki/2023_Mexico_City...  2023-10-27   
1098  17:00:00  https://en.wikipedia.org/wiki/2023_S%C3%A3o_Pa...  2023-11-03   
1099  06:00:00  https://en.wikipedia.org/wiki/2023_Las_Vegas_G...  2023-11-17   
1100  13:00:00  https://en.wikipedia.org/wiki/2023_Abu_Dhabi_G...  2023-11-24   

      fp1_time    fp2_date  fp2_time    fp3_date  fp3_time  quali_date  \
1096  17:30:00  2023-10-21  18:00:00          \N        \N  2023-10-20   
1097  18:30:00  2023-10-27  22:00:00  2023-10-28  17:30:00  2023-10-28   
1098  14:30:00  2023-11-04  14:30:00          \N        \N  2023-11-03   
1099  04:30:00  2023-11-17  08:00:00  2023-11-18  04:30:00  2023-11-18   
1100  09:30:00  2023-11-24  13:00:00  2023-11-25  10:30:00  2023-11-25   

     quali_time sprint_date sprint_time  
1096   21:00:00  2023-10-21    22:00:00  
1097   21:00:00          \N          \N  
1098   18:00:00  2023-11-04    18:30:00  
1099   08:00:00          \N          \N  
1100   14:00:00          \N          \N

In [101]:
# Elijo las columnas que son relevantes number el análisis
races_cols = races[["raceId","round","circuitId","name","date","time"]]

In [102]:
# Se revisa la dimensión, valores nulos, duplicados y los tipos de datos.
funciones.nudu(races_cols)

Dimensión del DF: (1101, 6) 

Valores nulos por columna:
raceId       0
round        0
circuitId    0
name         0
date         0
time         0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
raceId        int64
round         int64
circuitId     int64
name         object
date         object
time         object
dtype: object


In [103]:
races_cols['date_race'] = pd.to_datetime(races_cols['date'], format='%Y-%m-%d')
races_cols.drop('date', axis = 1, inplace = True)


C:\Users\SEBASTIAN\AppData\Local\Temp\ipykernel_13856\493878301.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  races_cols['date_race'] = pd.to_datetime(races_cols['date'], format='%Y-%m-%d')
C:\Users\SEBASTIAN\AppData\Local\Temp\ipykernel_13856\493878301.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  races_cols.drop('date', axis = 1, inplace = True)


In [104]:
funciones.nudu(races_cols)

Dimensión del DF: (1101, 6) 

Valores nulos por columna:
raceId       0
round        0
circuitId    0
name         0
time         0
date_race    0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
raceId                int64
round                 int64
circuitId             int64
name                 object
time                 object
date_race    datetime64[ns]
dtype: object


In [106]:
funciones.values_unique(races_cols)

Valores únicos en la columna raceId:
 [   1    2    3 ... 1118 1119 1120]

Valores únicos en la columna round:
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22]

Valores únicos en la columna circuitId:
 [ 1  2 17  3  4  6  5  9 20 11 12 13 14 15 22 18 24  7  8 10 16 19 21 70
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 71 73 76
 75 78 77 79 80]

Valores únicos en la columna name:
 ['Australian Grand Prix' 'Malaysian Grand Prix' 'Chinese Grand Prix'
 'Bahrain Grand Prix' 'Spanish Grand Prix' 'Monaco Grand Prix'
 'Turkish Grand Prix' 'British Grand Prix' 'German Grand Prix'
 'Hungarian Grand Prix' 'European Grand Prix' 'Belgian Grand Prix'
 'Italian Grand Prix' 'Singapore Grand Prix' 'Japanese Grand Prix'
 'Brazilian Grand Prix' 'Abu Dhabi Grand Prix' 'Canadian Grand Prix'
 'French Grand Prix' 'United States Grand Prix' 'San Marino Grand Prix'
 'Austrian Grand Prix' 'Argentine G

In [75]:
# Con el análisis anterior se observa que en la columna time hay valores con el valor \\N, se revisan estos valores:
round((len(races_cols[(races_cols.time == "\\N")])/len(races_cols))*100,2)

# Se observa que la cantidad de datos en la columna time que tienen como valor \\N representa un 66% del total
# Por esto se decide no tener en cuenta esta columna.

66.39

In [78]:
# Elijo las columnas que son relevantes number el análisis
races_cols = races_cols[["raceId","round","circuitId","name","date_race"]]

In [79]:
funciones.values_unique(races_cols)

Valores únicos en la columna raceId:
 [   1    2    3 ... 1118 1119 1120]

Valores únicos en la columna round:
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22]

Valores únicos en la columna circuitId:
 [ 1  2 17  3  4  6  5  9 20 11 12 13 14 15 22 18 24  7  8 10 16 19 21 70
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 71 73 76
 75 78 77 79 80]

Valores únicos en la columna name:
 ['Australian Grand Prix' 'Malaysian Grand Prix' 'Chinese Grand Prix'
 'Bahrain Grand Prix' 'Spanish Grand Prix' 'Monaco Grand Prix'
 'Turkish Grand Prix' 'British Grand Prix' 'German Grand Prix'
 'Hungarian Grand Prix' 'European Grand Prix' 'Belgian Grand Prix'
 'Italian Grand Prix' 'Singapore Grand Prix' 'Japanese Grand Prix'
 'Brazilian Grand Prix' 'Abu Dhabi Grand Prix' 'Canadian Grand Prix'
 'French Grand Prix' 'United States Grand Prix' 'San Marino Grand Prix'
 'Austrian Grand Prix' 'Argentine G

## Results

In [80]:
results.head()

resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  laps         time milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0         39    2       1:27.452         218.300         1  
1         41    3       1:27.739         217.586         1  
2         41    5       1:28.090         216.719         1  
3         58    7       1:28.603         215.464         1  
4         43    1       1:27.418         218.385         1

In [81]:
results.tail()

resultId  raceId  driverId  constructorId number  grid position  \
26075     26081    1110       817            213      3    19       16   
26076     26082    1110       858              3      2    18       17   
26077     26083    1110       807            210     27     0       18   
26078     26084    1110       832              6     55     4       \N   
26079     26085    1110       857              1     81     5       \N   

      positionText  positionOrder  points  laps       time milliseconds  \
26075           16             16     0.0    44  +1:43.071      5053521   
26076           17             17     0.0    44  +1:44.476      5054926   
26077           18             18     0.0    44  +1:50.450      5060900   
26078            R             19     0.0    23         \N           \N   
26079            R             20     0.0     0         \N           \N   

      fastestLap rank fastestLapTime fastestLapSpeed  statusId  
26075         25   15       1:50.994         227.169         1  
26076         37    9       1:50.486         228.213         1  
26077         26    4       1:49.907         229.415         1  
26078          9   19       1:53.138         222.864       130  
26079         \N    0             \N              \N       130

In [82]:
cant_N = round((len(results[results.time == "\\N"])/len(results))*100,2)
print(f"La cantidad de valores \\N en la columna time representa un {cant_N}%, por lo que no se tiene en cuenta")

La cantidad de valores \N en la columna time representa un 72.2%, por lo que no se tiene en cuenta


In [83]:
# Elijo las columnas que son relevantes number el análisis
results_cols = results[["resultId","raceId","driverId","constructorId","position","statusId"]]

In [84]:
# Se revisa la dimensión, valores nulos, duplicados y los tipos de datos.
funciones.nudu(results_cols)

Dimensión del DF: (26080, 6) 

Valores nulos por columna:
resultId         0
raceId           0
driverId         0
constructorId    0
position         0
statusId         0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
resultId          int64
raceId            int64
driverId          int64
constructorId     int64
position         object
statusId          int64
dtype: object


In [85]:
funciones.values_unique(results_cols)

Valores únicos en la columna resultId:
 [    1     2     3 ... 26083 26084 26085]

Valores únicos en la columna raceId:
 [  18   19   20 ... 1108 1109 1110]

Valores únicos en la columna driverId:
 [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  56
  63  62  59  66  54  55  57  58  60  61  64  65  68  69  70  71  72  73
  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90  91
  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108 110
 109 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127
 128 129 130 131 132 135 136 137 138 139 133 140 141 142 143 144 145 146
 147 148 151 149 150 152  67 153 154 155 156 157 158 159 163 160 161 162
 164 134 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198

In [86]:
results_cols["position"].replace("\\N","Retired", inplace = True)

C:\Users\SEBASTIAN\AppData\Local\Temp\ipykernel_13856\3894754750.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  results_cols["position"].replace("\\N","Retired", inplace = True)
C:\Users\SEBASTIAN\AppData\Local\Temp\ipykernel_13856\3894754750.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  results_cols["position"].replace("\\N","Retir

## Status

In [87]:
status.head()

statusId        status
0         1      Finished
1         2  Disqualified
2         3      Accident
3         4     Collision
4         5        Engine

In [88]:
status.tail()

statusId          status
134       137          Damage
135       138          Debris
136       139         Illness
137       140       Undertray
138       141  Cooling system

In [89]:
funciones.nudu(status)

Dimensión del DF: (139, 2) 

Valores nulos por columna:
statusId    0
status      0
dtype: int64

Valores duplicados: 0

Tipos de datos por columna:
statusId     int64
status      object
dtype: object


In [90]:
funciones.values_unique(status)

Valores únicos en la columna statusId:
 [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51 128  53  54
  55  56 129  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 130 131 132 133 134 135 136 137 138 139 140 141]

Valores únicos en la columna status:
 ['Finished' 'Disqualified' 'Accident' 'Collision' 'Engine' 'Gearbox'
 'Transmission' 'Clutch' 'Hydraulics' 'Electrical' '+1 Lap' '+2 Laps'
 '+3 Laps' '+4 Laps' '+5 Laps' '+6 Laps' '+7 Laps' '+8 Laps' '+9 Laps'
 'Spun off' 'Radiator' 'Suspension' 'Brakes' 'Differential' 'Overheating'
 'Mechanical' 'Tyre' 'Driver Seat' 'Puncture' 'Driveshaft' 'Retired'
 'F

La regla del 107 % se introdujo para la temporada 1996 y estuvo en vigor hasta la temporada 2002 La regla del 107 % es una regulación deportiva que afecta a las sesiones de clasificación de las carreras de Fórmula 1

# Unión

In [204]:
# circuits_cols: circuitId	circuitRef	location	country	lat	lng OK
# constructors_cols: constructorId	constructorRef	name	nationality OK
# drivers_cols: driverId	driverRef	forename	surname	birthdate	nationality OK
# lap_times_cols: raceId	driverId	lap	position	time 
# pit_stop_cols: raceId	driverId	stop	lap	duration
# races_cols: 	raceId	year	round	circuitId	name	date OK
# results_cols: 	resultId	raceId	driverId	constructorId	position	time	statusId OK 
# status: statusId	status OK

In [91]:
# Unión de las tablas
df_f1 = pd.merge(results_cols.iloc[:,1:], races_cols, on = "raceId").merge(constructors_cols, on = "constructorId", suffixes=('_race','_constructor')).merge(circuits_cols, on = "circuitId").merge(drivers_cols, on = "driverId", suffixes=('_constructor','_driver')).merge(status, on = "statusId")

In [92]:
# Nulos y duplicados
funciones.nudu(df_f1)

Dimensión del DF: (26080, 20) 

Valores nulos por columna:
raceId                     0
driverId                   0
constructorId              0
position                   0
statusId                   0
round                      0
circuitId                  0
name_race                  0
date_race                  0
name_constructor           0
nationality_constructor    0
location                   0
country                    0
lat                        0
lng                        0
forename                   0
surname                    0
birthdate                  0
nationality_driver         0
status                     0
dtype: int64

Valores duplicados: 3

Tipos de datos por columna:
raceId                              int64
driverId                            int64
constructorId                       int64
position                           object
statusId                            int64
round                               int64
circuitId                           int64
na

In [93]:
df_f1_dropcols = df_f1.copy()
for cols in df_f1_dropcols.columns:
    if "Id" in cols:
        df_f1_dropcols.drop(cols, axis = 1, inplace=True)

La regla del 107 % se introdujo para la temporada 1996 y estuvo en vigor hasta la temporada 2002

In [108]:
df_2002 = df_f1_dropcols[df_f1_dropcols.date_race.dt.year > 2002]

# Se guarda el archivo en formato PKL y parquet ya que permiten guardar cada columna en el formato definido en el df.

## Se usará el .parquet ya que se puede cargar en Power BI

In [112]:
df_2002.to_pickle('formula_1.pkl')
df_2002.to_parquet('formula_1.parquet', index = False)

In [113]:
pd.read_parquet('formula_1.parquet')

position  round              name_race  date_race name_constructor  \
0           1      1  Australian Grand Prix 2008-03-16          McLaren   
1           2      1  Australian Grand Prix 2008-03-16       BMW Sauber   
2           3      1  Australian Grand Prix 2008-03-16         Williams   
3           4      1  Australian Grand Prix 2008-03-16          Renault   
4           5      1  Australian Grand Prix 2008-03-16          McLaren   
...       ...    ...                    ...        ...              ...   
8286       16     12     Belgian Grand Prix 2023-07-30       AlphaTauri   
8287       17     12     Belgian Grand Prix 2023-07-30         Williams   
8288       18     12     Belgian Grand Prix 2023-07-30     Haas F1 Team   
8289  Retired     12     Belgian Grand Prix 2023-07-30          Ferrari   
8290  Retired     12     Belgian Grand Prix 2023-07-30          McLaren   

     nationality_constructor   location    country      lat        lng  \
0                    British  Melbourne  Australia -37.8497  144.96800   
1                     German  Melbourne  Australia -37.8497  144.96800   
2                    British  Melbourne  Australia -37.8497  144.96800   
3                     French  Melbourne  Australia -37.8497  144.96800   
4                    British  Melbourne  Australia -37.8497  144.96800   
...                      ...        ...        ...      ...        ...   
8286                 Italian        Spa    Belgium  50.4372    5.97139   
8287                 British        Spa    Belgium  50.4372    5.97139   
8288                American        Spa    Belgium  50.4372    5.97139   
8289                 Italian        Spa    Belgium  50.4372    5.97139   
8290                 British        Spa    Belgium  50.4372    5.97139   

      forename     surname  birthdate nationality_driver            status  
0        Lewis    Hamilton 1985-01-07            British          Finished  
1         Nick    Heidfeld 1977-05-10             German          Finished  
2         Nico     Rosberg 1985-06-27             German          Finished  
3     Fernando      Alonso 1981-07-29            Spanish          Finished  
4       Heikki  Kovalainen 1981-10-19            Finnish          Finished  
...        ...         ...        ...                ...               ...  
8286    Daniel   Ricciardo 1989-07-01         Australian          Finished  
8287     Logan    Sargeant 2000-12-31           American          Finished  
8288      Nico  Hülkenberg 1987-08-19             German          Finished  
8289    Carlos       Sainz 1994-09-01            Spanish  Collision damage  
8290     Oscar     Piastri 2001-04-06         Australian  Collision damage  

[8291 rows x 15 columns]